File: app/src/main/java/com/commonpepper/photosen/ui/adapters/PhotoAdapter.java
package com.commonpepper.photosen.ui.adapters;

import android.app.Activity;
import android.content.Intent;
import android.content.res.TypedArray;
import android.graphics.Color;
import android.graphics.drawable.GradientDrawable;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.ProgressBar;
import android.widget.TextView;

import com.commonpepper.photosen.R;
import com.commonpepper.photosen.network.NetworkState;
import com.commonpepper.photosen.network.datasource.AbstractListDataSource;
import com.commonpepper.photosen.network.model.Photo;
import com.commonpepper.photosen.ui.activities.SinglePhotoActivity;
import com.google.android.material.button.MaterialButton;
import com.squareup.picasso.Picasso;

import java.util.Random;

import androidx.annotation.NonNull;
import androidx.core.app.ActivityCompat;
import androidx.core.app.ActivityOptionsCompat;
import androidx.paging.PagedList;
import androidx.paging.PagedListAdapter;
import androidx.recyclerview.widget.RecyclerView;

public class PhotoAdapter extends PagedListAdapter<Photo, RecyclerView.ViewHolder> {
    private static final String TAG = PhotoAdapter.class.getSimpleName();

    private static final int VIEW_TYPE_PHOTO = 1;
    private static final int VIEW_TYPE_EXTRA = 2;

    private NetworkState networkState = NetworkState.RUNNING;

    public PhotoAdapter() {
        super(Photo.DIFF_CALLBACK);
    }

    @NonNull
    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        LayoutInflater layoutInflater = LayoutInflater.from(parent.getContext());
        if (viewType == VIEW_TYPE_PHOTO) {
            return new PhotoItemViewHolder(layoutInflater.inflate(R.layout.item_image, parent, false));
        } else {
            return new NetworkItemViewHolder(layoutInflater.inflate(R.layout.item_extra, parent, false));
        }
    }

    @Override
    public int getItemViewType(int position) {
        if (networkState != NetworkState.SUCCESS && position == getItemCount() - 1) {
            return VIEW_TYPE_EXTRA;
        } else {
            return VIEW_TYPE_PHOTO;
        }
    }

    @Override
    public int getItemCount() {
        if (hasExtraRow()) {
            return super.getItemCount() + 1;
        } else {
            return super.getItemCount();
        }
    }

    @Override
    public void onBindViewHolder(@NonNull RecyclerView.ViewHolder holder, int position) {
        if (holder instanceof PhotoItemViewHolder) {
            ((PhotoItemViewHolder) holder).bind(getItem(position));
        } else if (holder instanceof NetworkItemViewHolder) {
            ((NetworkItemViewHolder) holder).bind(networkState);
        }
    }


    private boolean hasExtraRow() {
        return networkState != null && networkState != NetworkState.SUCCESS;
    }

    public void setNetworkState(NetworkState newNetworkState) {
        NetworkState previousState = networkState;
        boolean hadExtraRow = hasExtraRow();
        networkState = newNetworkState;
        boolean hasExtraRow = hasExtraRow();
        if (hadExtraRow != hasExtraRow) {
            if (hadExtraRow) {
                notifyItemRemoved(getItemCount());
            } else {
                notifyItemInserted(getItemCount());
            }
        } else if (hasExtraRow && previousState != newNetworkState) {
            notifyItemChanged(getItemCount() - 1);
        }
    }

    private class PhotoItemViewHolder extends RecyclerView.ViewHolder {

        private ImageView mImageView;

        public PhotoItemViewHolder(@NonNull View itemView) {
            super(itemView);
            mImageView = itemView.findViewById(R.id.item_image_imageView);
        }

        public void bind(Photo photo) {
            GradientDrawable gd = new GradientDrawable();
            gd.setSize(photo.getWidth_z(), photo.getHeight_z());
            gd.setShape(GradientDrawable.RECTANGLE);
            TypedArray colors = mImageView.getResources().obtainTypedArray(R.array.scroll_colors);
            int index = (int) (Math.random() * colors .length());
            int color = colors.getColor(index, Color.BLACK);
            gd.setColor(color);
            colors.recycle();

            Picasso.get().load(photo.getUrl_z()).placeholder(gd).into(mImageView);

            mImageView.setOnClickListener(v -> {
                Intent intent = new Intent(mImageView.getContext(), SinglePhotoActivity.class);
                intent.putExtra(SinglePhotoActivity.PHOTO_TAG, photo);
                Bundle options = ActivityOptionsCompat.makeSceneTransitionAnimation((Activity)mImageView.getContext(),
                        mImageView, "sharedImageView").toBundle();
                ActivityCompat.startActivity(mImageView.getContext(), intent, options);
            });
        }
    }

    private class NetworkItemViewHolder extends RecyclerView.ViewHolder {

        private ProgressBar mProgressBar;
        private TextView mTextView;
        private MaterialButton mButton;

        public NetworkItemViewHolder(@NonNull View itemView) {
            super(itemView);
            mProgressBar = itemView.findViewById(R.id.item_last_progressBar);
            mTextView = itemView.findViewById(R.id.item_last_error_textView);
            mButton = itemView.findViewById(R.id.item_last_refreshButton);

            mButton.setOnClickListener(v -> {
                PagedList<Photo> list = getCurrentList();
                if (list != null) {
                    if (list.size() == 0) {
                        list.getDataSource().invalidate();
                    } else {
                        ((AbstractListDataSource) list.getDataSource()).retryLast();
                    }
                }
            });
        }

        public void bind(NetworkState networkState) {
            if (networkState == NetworkState.RUNNING) {
                mProgressBar.setVisibility(View.VISIBLE);
                mTextView.setVisibility(View.GONE);
                mButton.setVisibility(View.GONE);
            } else {
                mProgressBar.setVisibility(View.GONE);
                mTextView.setVisibility(View.VISIBLE);
                mButton.setVisibility(View.VISIBLE);
            }
        }
    }
}

File: app/src/main/java/com/commonpepper/photosen/ui/adapters/PhotoAdapter.kt
package com.commonpepper.photosen.ui.adapters

import android.app.Activity
import android.content.Intent
import android.graphics.Color
import android.graphics.drawable.GradientDrawable
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.ProgressBar
import android.widget.TextView
import androidx.core.app.ActivityCompat
import androidx.core.app.ActivityOptionsCompat
import androidx.paging.PagedListAdapter
import androidx.recyclerview.widget.RecyclerView.ViewHolder
import com.commonpepper.photosen.R.*
import com.commonpepper.photosen.model.Photo
import com.commonpepper.photosen.network.NetworkState
import com.commonpepper.photosen.network.datasource.AbstractListDataSource
import com.commonpepper.photosen.ui.activities.SinglePhotoActivity
import com.google.android.material.button.MaterialButton
import com.squareup.picasso.Picasso

class PhotoAdapter : PagedListAdapter<Photo?, ViewHolder>(Photo.DIFF_CALLBACK) {
    private var networkState: NetworkState? = NetworkState.RUNNING
    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val layoutInflater: LayoutInflater = LayoutInflater.from(parent.context)
        return if (viewType == VIEW_TYPE_PHOTO) {
            PhotoItemViewHolder(layoutInflater.inflate(layout.item_image, parent, false))
        } else {
            NetworkItemViewHolder(layoutInflater.inflate(layout.item_extra, parent, false))
        }
    }

    override fun getItemViewType(position: Int): Int {
        return if (networkState !== NetworkState.SUCCESS && position == itemCount - 1) {
            VIEW_TYPE_EXTRA
        } else {
            VIEW_TYPE_PHOTO
        }
    }

    override fun getItemCount(): Int {
        return if (hasExtraRow()) {
            super.getItemCount() + 1
        } else {
            super.getItemCount()
        }
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        if (holder is PhotoItemViewHolder) {
            holder.bind(getItem(position))
        } else if (holder is NetworkItemViewHolder) {
            holder.bind(networkState)
        }
    }

    private fun hasExtraRow(): Boolean {
        return networkState != null && networkState !== NetworkState.SUCCESS
    }

    fun setNetworkState(newNetworkState: NetworkState) {
        val previousState = networkState
        val hadExtraRow = hasExtraRow()
        networkState = newNetworkState
        val hasExtraRow = hasExtraRow()
        if (hadExtraRow != hasExtraRow) {
            if (hadExtraRow) {
                notifyItemRemoved(itemCount)
            } else {
                notifyItemInserted(itemCount)
            }
        } else if (hasExtraRow && previousState !== newNetworkState) {
            notifyItemChanged(itemCount - 1)
        }
    }

    private inner class PhotoItemViewHolder(itemView: View) : ViewHolder(itemView) {
        private val mImageView: ImageView = itemView.findViewById(id.item_image_imageView)
        fun bind(photo: Photo?) {
            val gd = GradientDrawable()
            gd.setSize(photo!!.width_z, photo.height_z)
            gd.shape = GradientDrawable.RECTANGLE
            val colors = mImageView.resources.obtainTypedArray(array.scroll_colors)
            val index = (Math.random() * colors.length()).toInt()
            val color = colors.getColor(index, Color.BLACK)
            gd.setColor(color)
            colors.recycle()
            if (!photo.isSexuallyExplicit) {
                Picasso.get().load(photo.url_z).placeholder(gd).into(mImageView)
                mImageView.setOnClickListener {
                    val intent = Intent(mImageView.context, SinglePhotoActivity::class.java)
                    intent.putExtra(SinglePhotoActivity.PHOTO_TAG, photo)
                    val options = ActivityOptionsCompat.makeSceneTransitionAnimation((mImageView.context as Activity),
                            mImageView, "sharedImageView").toBundle()
                    ActivityCompat.startActivity(mImageView.context, intent, options)
                }
            } else {
                mImageView.setImageDrawable(mImageView.resources.getDrawable(drawable.explicit_content))
                mImageView.setOnClickListener { v: View? -> }
            }
        }

    }

    private inner class NetworkItemViewHolder(itemView: View) : ViewHolder(itemView) {
        private val mProgressBar: ProgressBar = itemView.findViewById(id.item_last_progressBar)
        private val mTextView: TextView = itemView.findViewById(id.item_last_error_textView)
        private val mButton: MaterialButton = itemView.findViewById(id.item_last_refreshButton)
        fun bind(networkState: NetworkState?) {
            if (networkState === NetworkState.RUNNING) {
                mProgressBar.visibility = View.VISIBLE
                mTextView.visibility = View.GONE
                mButton.visibility = View.GONE
            } else {
                mProgressBar.visibility = View.GONE
                mTextView.visibility = View.VISIBLE
                mButton.visibility = View.VISIBLE
            }
        }

        init {
            mButton.setOnClickListener {
                val list = currentList
                if (list != null) {
                    if (list.size == 0) {
                        list.dataSource.invalidate()
                    } else {
                        (list.dataSource as AbstractListDataSource<*>).retryLast()
                    }
                }
            }
        }
    }

    companion object {
        private val TAG = PhotoAdapter::class.java.simpleName
        private const val VIEW_TYPE_PHOTO = 1
        private const val VIEW_TYPE_EXTRA = 2
    }
}